contracts/rewards/src/contract.rs

```diff
@@ -26,21 +29,90 @@ pub fn instantiate(
         },
     )?;
 
+    PARAMS.save(
+        deps.storage,
+        &StoredParams {
+            params: msg.params,
+            last_updated: Epoch {
+                epoch_num: 0,
+                block_height_started: env.block.height,
+            },
+        },
+    )?;
+
     Ok(Response::new())
 }
 
 #[cfg_attr(not(feature = "library"), entry_point)]
 pub fn execute(
     deps: DepsMut,
-    _env: Env,
-    _info: MessageInfo,
+    env: Env,
+    info: MessageInfo,
     msg: ExecuteMsg,
 ) -> Result<Response, axelar_wasm_std::ContractError> {
-    let mut _contract = Contract::new(deps);
     match msg {
-        ExecuteMsg::RecordParticipation { .. } => todo!(),
-        ExecuteMsg::AddRewards { .. } => todo!(),
-        ExecuteMsg::DistributeRewards { .. } => todo!(),
-        ExecuteMsg::UpdateParams { .. } => todo!(),
+        ExecuteMsg::RecordParticipation {
+            event_id,
+            worker_address,
+        } => {
+            let worker_address = deps.api.addr_validate(&worker_address)?;
+            Contract::new(deps)
+                .record_participation(
+                    nonempty::String::try_from(event_id)
+                        .change_context(ContractError::InvalidEventId)?,
+                    worker_address,
+                    info.sender,
+                    env.block.height,
+                )
+                .map_err(axelar_wasm_std::ContractError::from)?;
+
+            Ok(Response::new())
+        }
+        ExecuteMsg::AddRewards { contract_address } => {
+            let contract_address = deps.api.addr_validate(&contract_address)?;
+            let mut contract = Contract::new(deps);
+            let amount = info
+                .funds
+                .iter()
+                .find(|coin| coin.denom == contract.config.rewards_denom)
+                .filter(|_| info.funds.len() == 1)
```

Comment:
good idea. Since it's not obvious why this check exists and we (or I) will definitely forget what it's for, could you add an explanatory comment? something like "make sure the expected denomination is the only one attached to this message so other amounts don't get silently swallowed". No strong opinion on this, but for me it makes a bit more sense to first check that you only have one denomination and then check that it's the correct one.

Reply:
Agree. This `.filter` in the end is confusing.